File: examples/cpp/src/PublisherApp.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include <cameo/api/cameo.h>
#include <iostream>

using namespace cameo;

int main(int argc, char *argv[]) {

	// Initialize cameo.
	This::init(argc, argv);

	// Define the stop handler to properly stop.
	This::handleStop([] {});

	int numberOfSubscribers = 1;

	std::unique_ptr<coms::Publisher> publisher;

	try {
		// Create the publisher.
		publisher = coms::Publisher::create("the-publisher");
		publisher->setWaitForSubscribers(numberOfSubscribers);

		std::cout << "Created publisher " << *publisher << std::endl;

		publisher->init();
	}
	catch (const InitException& e) {
		std::cout << "Publisher error" << std::endl;
		return EXIT_FAILURE;
	}

	// Set the state.
	This::setRunning();

	std::cout << "Synchronized with " << numberOfSubscribers << " subscriber(s)" << std::endl;

	int i = 0;
	while (!This::isStopping()) {

		// Send a message.
		publisher->send(std::string{"a message "} + std::to_string(i));
		i++;
		
		// Sleep for 1s.
		std::this_thread::sleep_for(std::chrono::seconds(1));
	}

	std::cout << "Finished the application" << std::endl;

	This::terminate();

	return EXIT_SUCCESS;
}

File: examples/python/src/publisherapp.py
import sys
import cameopy
import time
import json

# Initialize cameo.
this = cameopy.This
this.init(sys.argv)

# Define the stop handler to properly stop.
def stop():
    print("Stop the app")

this.handleStop(stop)

numberOfSubscribers = 1

# Create the publisher.
publisher = cameopy.coms.Publisher.create("the-publisher")
publisher.setWaitForSubscribers(numberOfSubscribers)

print("Created publisher", publisher.getName())

publisher.init()

print("Synchronized with", numberOfSubscribers, "subscriber(s)")

# Set the state.
this.setRunning()

i = 0
while not this.isStopping():
    # Send a message.
    message = json.dumps({'message': "a message", 'value': i})
    publisher.send(message)
  
    i += 1
  
    time.sleep(1.000)

# Terminate the publisher and This.
publisher.terminate()
this.terminate()